### PR TITLE
[Refactor] Refactor top-down heatmap heads

### PR DESCRIPTION
**Pull request by liqikai9**
## Motivation

Refactor top-down heatmap heads.

## Modification

1. Add `mmpose/models/heads/heatmap_heads/cpm_head.py` for `CPM` and `Stacked Hourglass Networks`
Add `mmpose/models/heads/heatmap_heads/mspn_head.py` for `MSPN` and `RSN`.
Add `mmpose/models/heads/heatmap_heads/vipnas_head.py` for `ViPNAS`.

2. Modify `mmpose/datasets/pipelines/topdown_transforms.py` for generating multi-level heatmaps.

3. Add unittests for the above modules.

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit tests to ensure correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmpose/models/heads/heatmap_heads/multi_stage_heatmap_head.py [hunk 1]
```diff
@@ -0,0 +1,643 @@
...
+
+    .. _`Convolutional Pose Machines`: https://arxiv.org/abs/1602.00134
+    .. _`Stacked Hourglass Networks`: https://arxiv.org/abs/1603.06937
+    """
+
+    _version = 2
+
+    def __init__(self,
+                 num_stages: int,
+                 in_channels: Union[int, Sequence[int]],
```
Comment (ly015) [inline mmpose/models/heads/heatmap_heads/multi_stage_heatmap_head.py:57]: I suggest keep `in_channes` and `out_channels` as the first 2 parameters.

Comment (liqikai9) [reply mmpose/models/heads/heatmap_heads/multi_stage_heatmap_head.py:56]: Done

File: mmpose/models/heads/heatmap_heads/heatmap_head.py [hunk 2]
```diff
@@ -72,6 +76,7 @@ def __init__(self,
                  out_channels: int,
                  deconv_out_channels: OptIntSeq = (256, 256, 256),
                  deconv_kernel_sizes: OptIntSeq = (4, 4, 4),
+                 deconv_num_groups: OptIntSeq = (1, 1, 1),
```
Comment (ly015) [inline mmpose/models/heads/heatmap_heads/heatmap_head.py:79]: On second thought, I feel it's better to define a new class `ViPNASHead` head that inherits `HeatmapHead` and overrides the method `_make_deconv_layers`. So we don't have to make `HeatmapHead` more complex and can add a paper reference to the docstring of `ViPNASHead`.

File: mmpose/models/heads/heatmap_heads/multi_stage_heatmap_head.py [hunk 3]
```diff
@@ -0,0 +1,643 @@
...
+                # use the same loss_module over different stages
+                loss_func = self.loss_module
+
+            # the `gt_heatmaps` and `keypoint_weights` used to calculate loss
+            # for different stages are the same
+            loss_i = loss_func(multi_stage_pred_heatmaps[i], gt_heatmaps,
+                               keypoint_weights)
+
+            if 'loss_heatmap' not in losses:
+                losses['loss_heatmap'] = loss_i
```
Comment (ly015) [inline mmpose/models/heads/heatmap_heads/multi_stage_heatmap_head.py:262]: I suggest using "loss_kpt"

File: mmpose/models/heads/heatmap_heads/multi_stage_heatmap_head.py [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+        feature = self.conv_layers(feature)
+        output = nn.functional.interpolate(
+            feature, size=self.out_shape, mode='bilinear', align_corners=True)
+        if self.use_prm:
+            output = self.prm(output)
+        return output
+
+
+@MODELS.register_module()
+class MultiStageMultiUnitHeatmapHead(BaseHead):
```
Comment (ly015) [inline mmpose/models/heads/heatmap_heads/multi_stage_heatmap_head.py:424]: What about placing `MultiStageMultiUnitHeatmapHead` in a separate file, e.g. `msmu_heatmap_head.py`?

File: mmpose/models/heads/heatmap_heads/heatmap_head.py (hunk 2 above)
Comment (liqikai9) [reply mmpose/models/heads/heatmap_heads/heatmap_head.py]: Changed.

File: mmpose/models/heads/heatmap_heads/multi_stage_heatmap_head.py (hunk 3 above)
Comment (liqikai9) [reply mmpose/models/heads/heatmap_heads/multi_stage_heatmap_head.py]: Changed.

File: mmpose/models/heads/heatmap_heads/multi_stage_heatmap_head.py (hunk 4 above)
Comment (liqikai9) [reply mmpose/models/heads/heatmap_heads/multi_stage_heatmap_head.py]: Good idea. Changed.

File: checkpoints [hunk 5]
```diff
@@ -0,0 +1 @@
+/home/PJLAB/liqikai/checkpoints/
```
Comment (jin-s13) [inline checkpoints:1]: unnecessary I suppose

Comment (liqikai9) [reply checkpoints:1]: You are right. It is a mistake. Will fix it.

**Review by ly015 — APPROVED** [review]

**Pull request merged**